В данной задаче вам предстоит решить задачу классификации изображений – отделить изображения кошек от изображений собак, используя классификацию с мягким зазором и параметром C. В предложенном архиве находится выборка, включающая в себя изображения кошек и собак (по 500 изображений). Имя каждого изображения, для удобства, имеет следующий формат:  cat/dog.номер_изображения.jpg  в зависимости от того, какое животное присутствует на изображении.
Обучите классификатор LinearSVC() при C = 0.65 и random_state = 12 на предложенной выборке. Используйте параметр random_state = 12 для разделения выборки на тренировочную и тестовую части, в отношении 75/25 функции train_test_split().

In [ ]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import StackingClassifier
import cv2
from imutils import paths
from sklearn.metrics import accuracy_score, precision_score, recall_score
from sklearn.metrics import confusion_matrix
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import train_test_split
import numpy as np

In [ ]:
!unzip -q /content/train_task.zip -d train
!unzip -q /content/test_task.zip -d test

replace train/train/.DS_Store? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace train/train/cat.0.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: a
error:  invalid response [a]
replace train/train/cat.0.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
replace test/test/.DS_Store? [y]es, [n]o, [A]ll, [N]one, [r]ename: y
replace test/test/cat.1000.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: y A
replace test/test/cat.1001.jpg? [y]es, [n]o, [A]ll, [N]one, [r]ename: A


In [ ]:
def extract_histogram(image, bins=(8, 8, 8)):
    hist = cv2.calcHist([image], [0, 1, 2], None, bins, [0, 256, 0, 256, 0, 256])
    cv2.normalize(hist, hist)
    return hist.flatten()

In [ ]:
X = []
y = []
imagePaths = sorted(list(paths.list_images('train')))
for image in imagePaths:
  image1 = cv2.imread(image)
  X.append(extract_histogram(image1))
  y.append(1) if 'cat' in image else y.append(0)

In [ ]:
X_test = []
y_test = []
imagePaths = sorted(list(paths.list_images('test')))
for image in imagePaths:
    image1 = cv2.imread(image)
    X_test.append(extract_histogram(image1))
    y_test.append(1) if 'cat' in image else y_test.append(0)

In [ ]:
X_train, y_train, X_val, y_val = train_test_split(X, y, test_size=0.25, random_state=12)

In [ ]:
linearSVC = LinearSVC(C = 0.65, random_state=12).fit(X, y)

In [ ]:
print(linearSVC.score(X, y))

0.713


In [ ]:
linearSVC.coef_[0][385]

0.0017285412244758902

In [ ]:
linearSVC.coef_[0][476]

-0.000563731016005073

In [ ]:
linearSVC.coef_[0][167]

-0.04395105555123137

In [ ]:
ny_pred = linearSVC.predict(X_test)

In [ ]:
accuracy_score(y_test, y_pred)

0.54

In [ ]:
precision = precision_score(y_test, y_pred)
precision

0.5294117647058824

In [ ]:
recall = recall_score(y_test, y_pred)
recall

0.72

In [ ]:
F = 2 * (precision * recall)/(precision + recall)
F

0.6101694915254237

In [ ]:
X_control = []
image1 = cv2.imread('/content/test/test/cat.1041.jpg')
X_control.append(extract_histogram(image1))
1 - linearSVC.predict(X_control)

array([1])

Введите значение коэффициента θ_386 построенной гиперплоскости: 0.00172854122447589

Введите значение коэффициента θ_476 построенной гиперплоскости: -0.000563731016005073

Введите значение коэффициента θ_168 построенной гиперплоскости: -0.04395105555123137

F1: 0.6101694915254237